## ML Zoomcamp HW Week 1

#### Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Q1. What is the version of Pandas installed?

In [2]:
pd.__version__

'2.0.3'

#### Getting the data from url and reading as a Dataframe

In [4]:
url = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv"

df = pd.read_csv(url, sep = ",")
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

Q2. How many columns are in the dataset?

- 10
- 6560
- 10989
- 20640

In [5]:
no_of_columns = len(df.columns)

In [6]:
no_of_columns

10

Q3. Which columns in the dataset have missing values?

- total_rooms
- total_bedrooms
- both of the above
- no empty columns in the dataset

In [7]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

The column total_bedrooms has missing values. A total of 207.

Q4. How many unique values does the ocean_proximity column have?

- 3
- 5
- 7
- 9

In [8]:
df.ocean_proximity.nunique()

5

There are 5 unique values in the column ocean_proximity. They are:

In [9]:
df.ocean_proximity.unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

Q5. What's the average value of the median_house_value for the houses located near the bay?

- 49433
- 124805
- 259212
- 380440

In [13]:
df[df['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean()

259212.31179039303

The average value of median_house_value for houses near the bay is approximately 259212.

Q6. 

1. Calculate the average of total_bedrooms column in the dataset.
2. Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
3. Now, calculate the average of total_bedrooms again.
4. Has it changed?

Has it changed?

(__Hint__: take into account only 3 digits after the decimal point.)

- Yes
- No

In [17]:
mean_val = df['total_bedrooms'].mean()

In [20]:
mean_val

537.8705525375618

In [18]:
df_copy = df.copy()
df_copy['total_bedrooms'].fillna(value = mean_val, inplace=True)

In [19]:
df_copy['total_bedrooms'].mean()

537.8705525375617

No, the average value has not changed till 3 decimal places.

Q7.

1. Select all the options located on islands.
2. Select only columns housing_median_age, total_rooms, total_bedrooms.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
5. Compute the inverse of XTX.
6. Create an array y with values [950, 1300, 800, 1000, 1300].
7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
8. What's the value of the last element of w?

(__Note__: You just implemented linear regression. We'll talk about it in the next lesson.)

- -1.4812
- 0.001
- 5.6992
- 23.1233

In [29]:
df_island = df[df['ocean_proximity'] == 'ISLAND'].copy()

In [31]:
df_island_selective = df_island[['housing_median_age', 'total_rooms', 'total_bedrooms']]
df_island_selective

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [32]:
X = df_island_selective.to_numpy()
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [34]:
XTX = X.T @ X
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [35]:
XTX_inverse = np.linalg.inv(XTX)
XTX_inverse

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [36]:
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [41]:
w = XTX_inverse @ X.T
w = w @ y
w

array([23.12330961, -1.48124183,  5.69922946])

The last element is 5.6992